### PR TITLE
Support larger datasets for question answering 

### PR DESCRIPTION
**Pull request by Zhilin123**
# What does this PR do ?

Support larger datasets for question answering 

**Collection**: NLP

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Review by yzhang123 — CHANGES_REQUESTED** [review]

File: examples/nlp/intent_slot_classification/data/assistant_utils.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
```
[the remark this answers is not in the document]
Comment (Zhilin123) [reply examples/nlp/intent_slot_classification/data/assistant_utils.py:1]: Sure!

File: examples/nlp/intent_slot_classification/data/assistant_utils.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+    if not os.path.exists(our_infold + "/testset"):
+        shutil.copytree(old_infold + '/testset/csv', our_infold + '/testset')
+
+
+def get_intents(infold):
+    """ Get list of intents from file names. """
+    intents = [f[:-4] for f in os.listdir(infold)]
+    intents.sort()
+    print(f'Found {len(intents)} intents')
```
[the remark this answers is not in the document]
Comment (Zhilin123) [reply examples/nlp/intent_slot_classification/data/assistant_utils.py:48]: sure

File: examples/nlp/intent_slot_classification/data/assistant_utils.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+import os
+import re
+import shutil
+
+from nemo.collections.nlp.data.data_utils.data_preprocessing import DATABASE_EXISTS_TMP, if_exist, write_files
+from nemo.utils import logging
+
+
+def copy_input_files(infold):
+    """ Put training files in convenient place for conversion to our format. """
```
[the remark this answers is not in the document]
Comment (Zhilin123) [reply examples/nlp/intent_slot_classification/data/assistant_utils.py:24]: Sure

File: examples/nlp/intent_slot_classification/data/assistant_utils.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+    our_infold = infold + "/dataset"
```
[the remark this answers is not in the document]
Comment (Zhilin123) [reply examples/nlp/intent_slot_classification/data/assistant_utils.py:25]: This copying of files into a unified dir should have been designed for downstream functions to locate all relevant files efficiently.  This is mainly used on the assistant dataset (10s of MB) so space/time from rearranging the files might have been ok as convenience was the original intention (the initial external directory was rather messy).

File: examples/nlp/intent_slot_classification/data/assistant_utils.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+            phrases = query.split(";")
+            intent_query = phrases[4][1:-1] + "\t" + str(index)
+            intent_queries.append(intent_query)
+
+    return intent_queries
+
+
+def get_slots(infold, modes):
+    """
+    Find a lost of unique slot types in training and testing data.
```
[the remark this answers is not in the document]
Comment (Zhilin123) [reply examples/nlp/intent_slot_classification/data/assistant_utils.py:68]: sure

File: examples/nlp/intent_slot_classification/data/assistant_utils.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+    We use a single slot type name both for starting and continuation tokes (not using B-, I- notation).
```
[the remark this answers is not in the document]
Comment (Zhilin123) [reply examples/nlp/intent_slot_classification/data/assistant_utils.py:69]: sure

File: examples/nlp/intent_slot_classification/data/assistant_utils.py [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+            for line in lines[1:]:
+                query = line.split(";")[3]
+                slot_phrases = re.findall('\[.*?\]', query)
+                for slot_phrase in slot_phrases:
+                    slot = slot_phrase.split(" : ")[0][1:]
+                    slots.add(slot)
+
+    slots = sorted(slots)
+    slots.append("O")
+    print(f'Found {len(slots)} slot types')
```
[the remark this answers is not in the document]
Comment (Zhilin123) [reply examples/nlp/intent_slot_classification/data/assistant_utils.py:86]: sure updated

File: examples/nlp/intent_slot_classification/data/assistant_utils.py [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+    return slots
+
+
+def get_slot_queries(infold, slot_dict, mode, intent_names):
+    """ Convert each word in a query to corresponding slot number. """
```
[the remark this answers is not in the document]
Comment (Zhilin123) [reply examples/nlp/intent_slot_classification/data/assistant_utils.py:91]: sure

File: examples/nlp/intent_slot_classification/data/import_datasets.py [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+    if_exist,
+)
+from nemo.utils import logging
+
+
+def ids2text(ids, vocab):
+    return ' '.join([vocab[int(id_)] for id_ in ids])
+
+
+def process_atis(infold, outfold, modes=['train', 'test'], do_lower_case=False):
```
[the remark this answers is not in the document]
Comment (Zhilin123) [reply examples/nlp/intent_slot_classification/data/import_datasets.py:37]: sure

File: tutorials/nlp/Dialogue.ipynb [hunk 10]
```diff
@@ -28,7 +28,7 @@
       "source": [
         "import os \n",
         "!apt-get update && apt-get install -y libsndfile1 ffmpeg\n",
-        "!git clone https://github.com/NVIDIA/NeMo --branch main\n",
+        "!git clone https://github.com/NVIDIA/NeMo --branch r1.9.0\n",
```
[the remark this answers is not in the document]
Comment (Zhilin123) [reply tutorials/nlp/Dialogue.ipynb:31]: I think maybe r1.9.0 might be best for now as future stable releases might have code/dependency changes that accidentally render this tutorial not able to be used.

File: nemo/collections/nlp/models/question_answering/qa_model.py [hunk 11]
```diff
@@ -58,19 +57,21 @@ def __init__(self, cfg: DictConfig, trainer: Trainer = None):
...
     @typecheck()
     def forward(self, input_ids, attention_mask, token_type_ids):
-        hidden_states = self.bert_model(
-            input_ids=input_ids, token_type_ids=token_type_ids, attention_mask=attention_mask
-        )
-        if isinstance(hidden_states, tuple):
-            hidden_states = hidden_states[0]
-        logits = self.classifier(hidden_states=hidden_states)
-        return logits.float()
+        with autocast():
```
[the remark this answers is not in the document]
Comment (Zhilin123) [reply nemo/collections/nlp/models/question_answering/qa_model.py:60]: sure

**Review by yzhang123 — APPROVED** [review]

**Pull request merged**